---
title: "hw02"
author: "Eva Y"
date: "March 21, 2015"
output:
  html_document:
    keep_md: yes
---

## Attach packages
```{r, message=FALSE, warning=FALSE}
library(car)
library(edgeR)
library(gplots)
library(ggplot2)
library(GGally)
library(lattice)
library(limma)
library(RColorBrewer)
library(reshape2)
library(VennDiagram)
library(yeast2.db)
```

## Useful functions
After going through hw02, I realized there were many occasions that required repetition of the same analysis on different datasets. For this purpose, it is a good idea to functionalize some commonly used scripts.

```{r}
# 01. plot heatmap of Pearson's correlation between samples
pcorr <- function(data){
  # compute Pearson's correlation between samples
  pcorr_dat <- cor(data)
  diag(pcorr_dat) <- NA
  
  # plot heatmap
  heatmap.2(pcorr_dat, col=spectral.palette,
          trace="none",
          ColSideColors=sample.cols[unclass(sample.name$sample)], 
          RowSideColors=sample.cols[unclass(sample.name$sample)],
          margin=c(4,10))
  
  # add legend
  legend("topright", legend=levels(sample.name$sample),
       col=sample.cols, ncol=2, border=FALSE,
       lty=1, lwd=5, cex=0.5)
}

# 02. perform limma DEA and package output
do.limma.dea <- function(data, array=FALSE){
  # fit the linear model
  fit <- lmFit(data, design)
  
  # apply eBayes() to moderate the estimated error variances
  ebfit <- eBayes(fit)
  
  # get output of the linear model
  top <- topTable(ebfit, number=Inf)
  
  # if array, get gene.id
  if (array){
    # retrive gene IDs from yeast2ORF
    probe.id <- row.names(top)
    gene.id <- unlist(mget(probe.id, yeast2ORF))
    
    # make data frame
    data.frame(probe.id=probe.id, gene.id=gene.id,
                     p.value=top$P.Value, q.value=top$adj.P.Val,
                     log.fc=top$logFC, test.stat=top$t)
    } else {
      data.frame(gene.id=row.names(top),
                 p.value=top$P.Value, q.value=top$adj.P.Val,
                 log.fc=top$logFC, test.stat=top$t)
    }
  }

# 03. when a list of probe IDs and gene IDs are provided
#     get gene expression from array data and read counts from RNA seq data
#     generate a tall-and-skinny data frame 
get.genes <- function(probe.list, gene.list, x=""){
  # get gene expression from array data
  gene.exp <- subset(array.dat.fix, 
                     row.names(array.dat.fix) %in% probe.list)
  
  # make data frame tall and skinny
  df <- melt(t(gene.exp))
  colnames(df)[1:3] <- c("sample", "probe.id", "array")
  df <- data.frame(gene.id=rep(unlist(gene.list), each=6), df)
  
  # substitute sample labels with growth condition
  df$sample <- sub("b\\d", "batch", df$sample)
  df$sample <- sub("c\\d", "chemostat", df$sample)
  
  # get read counts from RNA seq data
  read.counts <- melt(t(subset(deep.count.dat, 
                          row.names(deep.count.dat) %in% gene.list)))$value
 
  # add read counts to the array data frame
  df <- data.frame(df, voomlimma=read.counts)
  
  # make data frame tall and skinny and add DEA status
  # DEA status (e.g. common.interesting = DE genes in both array and limma analysis)
  df <- melt(df, id.vars=c("sample", "probe.id", "gene.id"), 
                   measure.vars=c("array", "voomlimma"),
                   variable.name="DEA")
  df <- data.frame(df, DEA.status=x)
}

# 04. perform edgeR DEA and package output
do.edgeR.dea <- function(data){
  # produce an object of type DGEList 
  dge.glm <- DGEList(counts=data, group=metadata$condition)
  
  # estimate dispersion parameters
  com.disp <- estimateGLMCommonDisp(dge.glm, design, verbose=TRUE)
  trend.disp <- estimateGLMTrendedDisp(com.disp, design)
  tag.disp <- estimateGLMTagwiseDisp(trend.disp, design)
  
  # fit the model
  fit <- glmFit(tag.disp, design)
  lrt <- glmLRT(fit, coef=2)
  tt.glm <- topTags(lrt, n=Inf)
  
  # package results to the specified data frame
  data.frame(gene.id=row.names(tt.glm),
             p.value=tt.glm$table$PValue, q.value=tt.glm$table$FDR,
             log.fc=tt.glm$table$logFC, test.stat=tt.glm$table$LR)
}
```

## Q1 Microarray analysis

### Q1a) (0 points) Load Microarray Data

> Load the normalized data.

```{r}
# load normalized microarray data
array.dat <- read.table("../../data/GSE37599-data.tsv", header=TRUE, row.names=1)

# what is the flavour of the dataset? 
str(array.dat)
```

> What are dimensions of the dataset?

```{r}
dim(array.dat)
```

> In addition to reporting number of rows and columns, make it clear what rows and columns represent and how you're interpreting column names.**

```{r}
# what are the column names?
colnames(array.dat)

# what are the row names?
head(row.names(array.dat))
```

Based on this exploratory analysis, this dataset consists of 10928 rows and 6 columns. The rows represent the probes measured by the microarray platform whereas the columns represent yeast samples in the experiment. Three samples are labelled with "b" followed by 1-3. These are yeast samples grown in batch medium and the number following the letter represents the sample replicates. There are also three samples labelled with "c" followed by 1-3. These are yeast samples grown in the chemostat setup and the number following the letter represents the sample replicates. There are three replicates for each growth condition(total samples = 6).

### Q1b) (1 point) Identify Sample Swap

> The labels on two of the samples have been swapped, that is one of the batch samples has been labelled as chemostat and vice-versa. Produce the plots described below and explain how they allow you to identify the swapped samples.

> i. (High volume) scatter plot matrix.

```{r}
ggpairs(array.dat) + theme_bw()
```

From the far-left column of the scatter plot matrix, we can be observed that b1 is more highly correlated with c1 and c3 compared to b2 and b3. From the second last row, we can be observed that c2 is more highly correlated to b2 and b3 compared to c1. From the correlation plot between c2 and c3 (last row, second plot from the right), we can also observed that c2 is less correlated to c3. 

This indicates a sample swap between b1 and c2! 

> ii. A heatmap of the first 100 genes (you can try more but it gets slow).

```{r, message=FALSE, warning=FALSE}
# create data frame for heatmap.2
array.100 <- as.matrix(array.dat[1:100, ])

# extract sample names
sample.name <- data.frame(sample=colnames(array.100))

# color palette 
j.spectral <- colorRampPalette(brewer.pal(n=9, "Spectral"))
palette.size <- 256
spectral.palette <- j.spectral(palette.size)

# specify color labels for different samples
sample.cols <- brewer.pal(11, "RdGy")[c(2,3,4,8,9,10)]

# plot heatmap without sample clustering
heatmap.2(array.100, col=spectral.palette, 
          trace="none", Colv=NA, 
          ColSideColors=sample.cols[unclass(sample.name$sample)],
          main="Heatmap of 100 genes",
          margin=c(4,10))

# add legend
legend("topright", legend = levels(sample.name$sample),
       col=sample.cols, ncol=2, lty=1, lwd=5, cex=0.5)
```

From this heatmap without sample clustering (turned off column clustering), we can clearly see that the expression of certain genes are different between b1 and the other b samples. In addition, we can see that the expression of the same set of genes are differentially expressed between c2 and the other c samples. To point out an example, we can notice the darker blue line from the far left column for the b1 sample and this darker blue line is followed by red for b2 and b3. This is similarly observed for the same set of genes for the c samples, in which the darker blue line continued for c1, switched to red for c2, then switched back to darker blue for c3. 

This observation supports the sample swap between b1 and c2. 

Let's allow clustering for the samples. We should expect that b1 will cluster with c1 and c3 whereas c2 will cluster with b2 ad b3. 
```{r}
# plot heatmap with sample clustering
heatmap.2(array.100, col=spectral.palette, 
          trace="none", 
          ColSideColors=sample.cols[unclass(sample.name$sample)],
          main="Heatmap of 100 genes",
          margin=c(4,10))

# add legend
legend("topright", legend = levels(sample.name$sample),
       col=sample.cols, ncol=2, lty=1, lwd=5, cex=0.5)
```

As expected, we can observed that c2 clusters with b2 and b3 whereas b1 clusters with c1 and c3. 

> iii. Compute the Pearson correlation of the samples and plot the results using a heatmap.

```{r}
# use the pcorr() function for this question
pcorr(array.dat)
```

The heatmap showing Pearson's correlation between samples also demonstrated that c2 clusters with b2 and b3 while b1 clusters with c1 and c3. 

> iv. Scatterplot the six data samples with respect to the first two principal components and label the samples.

```{r}
# perform pca
pca <- prcomp(array.dat)

# extract PC1 and PC2
pca.df <- data.frame(pca$rotation)[ ,1:2]

# make data frame for plotting
pca.df <- data.frame(condition=rep(c("batch", "chemostat"), each=3), pca.df)

# plot pca with respect to the first two principal components
ggplot(pca.df, aes(x=PC1, y=PC2, color=condition, label=rownames(pca.df))) +
  geom_text() + 
  theme_bw()
```

Lastly, the scatterplot of the first two principal components showed that b1 clusters with c1 and c3 whereas c2 clusters with b2 and b3. 

In conclusion, all four plots showed that a sample swap occurs between b1 and c2.

### Q1c) (2 points) Microarray Differential Expression

>  Fix the label swap identified in question 1b. We want to swap b1 <--> c2. Revisit one or more elements of question 1b to sanity check before proceeding.

Now to fix the label swap and some sanity check before proceeding.
```{r}
# fix label swap
array.dat.fix <- with(array.dat, data.frame(row.names=row.names(array.dat), 
                                  b1=array.dat$c2, b2=array.dat$b2, 
                                  b3=array.dat$b3, c1=array.dat$c1, 
                                  c2=array.dat$b1, c3=array.dat$c3))
```

Time to revisit one or more elements of question 1b to sanity check before proceeding...I pick the heatmap showing Pearson's correlation between samples. 
```{r}
pcorr(array.dat.fix)
```

The samples are clustered in the correct groups. Unlike the previous heatmap depicting Pearson's correlation between samples, we can see that the grey labels (chemostat growth condition) are clustered together while red labels (batch medium) are clustered together.

Everything checked out. Good to go!

> Now use this data to do a differential expression analysis with `limma`.

> Package these results in a data frame with six columns:
  - probe.id - The array probe id.
  - gene.id - The id of the gene which the probe overlaps (see below).
  - p.value - The raw p-value for the probe.
  - q.value - The BH corrected p-value, aka the q-value.
  - log.fc - The log fold change which is the column called "logFC" in the limma results table.
  - test.stat - The test statistics which for limma is the moderated t statistic. This is the column called "t" in the limma results table.
  
First, we need to generate a design matrix before fitting the linear model. 
```{r}
# create metadata
metadata <- data.frame(sample=colnames(array.dat.fix), 
                       condition=rep(c("batch", "chemostat"), each=3))

# create design matrix
design <- model.matrix(~condition, metadata)
design
```

Using the `do.limma.dea()` function, we can perform DEA and package the data as specified above.
```{r, message=FALSE}
limma.array.results <- do.limma.dea(array.dat.fix, array=TRUE)

# check dimensions of the dataframe
str(limma.array.results, max.level=0)
```

> Remove any rows with probes which don't map to genes. You'll be able to find these because they will have NA as their gene id. Work with this data.frame to answer the questions below.

Based on the description above, let's filter out rows with NA as their gene IDs.
```{r}
# filter out NA values
limma.filter.array.results <- subset(limma.array.results, 
                                     limma.array.results$gene.id != "NA")

# check dimensions of the filtered dataset
str(limma.filter.array.results, max.level=0)
```

> i. How many probes did we start with and how many remain after removing probes without gene ids?

```{r}
# pre-filtered datasets 
nrow(limma.array.results)

# dataset without NA values as gene IDs
nrow(limma.filter.array.results)
```

We started with `r nrow(limma.array.results)` probes and we are left with `r nrow(limma.filter.array.results)` after filtering out gene IDs with NA values. 

> ii. Illustrate the differential expression between the batch and the chemostat samples for the top hit (i.e., probe with the lowest p- or q-value).

```{r, message=FALSE}
# top hit with lowest q-value
top.hit <- limma.filter.array.results$probe.id[1]

# select gene expression for the probe for all samples from array.dat.fix
top.hit <- array.dat.fix[top.hit, ]

# make data frame tall and skinny
top.hit <- melt(top.hit)
top.hit <- cbind(gene=rep(limma.filter.array.results$probe.id[1], 
                          each=nrow(top.hit)), 
                 condition=metadata$condition, top.hit)

# plot 
ggplot(top.hit, aes(x=condition, y=value)) + 
  geom_point() +
  facet_wrap(~gene) +
  theme_bw() +
  stat_summary(aes(group=1), fun.y=mean, geom="line") +
  ylab("Gene expression") +
  xlab("Growth condition") 
```

> iii. How many probes are identified as differentially expressed at a false discovery rate (FDR) of 1e-05 (note: this is a FDR cutoff used in the original paper)?

```{r}
# number of probes with FDR < 1e-05
nrow(subset(limma.filter.array.results, q.value < 1e-05))

# save differentially expressed genes with FDR < 1e-05
array.dea.fdr <- subset(limma.filter.array.results, q.value < 1e-05)
```

At FDR < 1e-05, there are `r nrow(subset(limma.filter.array.results, q.value < 1e-05))` differentially expressed genes between yeast samples grown under batch medium and chemostat setup. 

> iv. Save your results for later with `write.table()`.

```{r}
# save DEA results
write.table(limma.filter.array.results, "array.limma.results.tsv", row.names=TRUE, col.names=NA)
```


## Q2 RNA-Seq Analysis

### Q2a) (1 point) Load RNA Count Data and Sanity Check

> Load the count data. 

```{r}
# load count data
deep.count.dat <- read.table("../../data/stampy.deep.counts.tsv", header=TRUE, row.names=1)

# what is the flavour of the dataset? 
str(deep.count.dat)
```

> i) What are dimensions of the dataset? In addition to reporting number of rows and columns, make it clear what rows and columns represent. What is the difference between the rows of this dataset versus rows of the array data in question 1a?

```{r}
# dimensions of the dataset
dim(deep.count.dat)

# what do the columns represent? 
colnames(deep.count.dat)

# what do the rows represent?
head(row.names(deep.count.dat))
```

The count dataset has `r nrow(deep.count.dat)` rows and `r ncol(deep.count.dat)` columns. The columns represent the 6 yeast samples grown under batch medium (b group) and chemostat setup (c group). Unlike the rows in the microarray data which represent probe IDs, the rows in the count data represent sequence annotations of the yeast reference genome. While the numeric values in the microarray data are gene expression, the numeric values in the count data are number of reads. 

> ii) Do a sanity check to make sure there is no sample swap by plotting a heatmap of the sample correlations.

```{r}
# use pcorr() function to do this
pcorr(deep.count.dat)
```

Grey labels (c group) are clustered together whereas red labels (b group) are clustered together. No sample swap is identified, we can move on.

### Q2b) (2 points) voom Differential Expression Analysis

> Use `voom+limma` to identify differentially expressed genes between the batch medium vs. chemostat conditions.

> i) `voom` normalizes the counts before it converts counts to log2-cpm. Use calcNormFactors to normalize counts.

```{r}
# normalize counts
norm.factor <- calcNormFactors(deep.count.dat)
```

> ii) Use `voom` to convert count data into logged CPM data and then use limma to identify differentially expressed genes between conditions.

> Package these results in a data.frame called 'voom.limma.deep.results' with five columns:
  - gene.id - The id of the gene which reads were aligned to.
  - p.value - The raw p-value for the gene.
  - q.value - The BH corrected p-value, aka the q-value.
  - log.fc - The log fold change which is the column called "logFC" in the edgeR results table.
  - test.stat - The test statistic, which is the column called "t".

```{r, message=FALSE}
# convert count data to logged CPM using voom()
deep.count.norm <- voom(deep.count.dat, design, plot=TRUE,
                  lib.size=colSums(deep.count.dat)*norm.factor)

# again, we can utilize the do.limma.dea() function
voom.limma.deep.results <- do.limma.dea(deep.count.norm, array=FALSE)
```

> iii) How many genes are differentially expressed between conditions at a false discovery rate (FDR) of 1e-05?

```{r}
# number of genes with FDR < 1e-05
nrow(subset(voom.limma.deep.results, q.value < 1e-05))

# save differentially expressed genes with FDR < 1e-05
deep.dea.fdr <- subset(voom.limma.deep.results, q.value < 1e-05)
```

There are `r nrow(subset(voom.limma.deep.results, q.value < 1e-05))` genes that are differentially expressed between conditions at FDR < 1e-05. 

> iv) Save your results for later with `write.table()` in file called stampy.deep.limma.results.tsv.

```{r}
# save DEA results
write.table(voom.limma.deep.results, "stampy.deep.limma.results.tsv", 
            row.names=TRUE, col.names=NA)
```


## Q3) Compare DEA results between RNA-Seq and array

### Q3a) (1 point) Comparing volumes of DE genes

> In this question, you will examine the difference between the q-values from both analyses (i.e., array and voom+limma) by overlaying density plots of the q-values from each analysis.

We'll be working with these two datasets for this question.
```{r}
# limma DEA results for microarray data
str(limma.filter.array.results)
  
# voom+limma DEA results for deep count data
str(voom.limma.deep.results)
```

Generate density plot that represents q.values of common genes analyzed by both platforms.
```{r, message=FALSE}
# what are the genes analyzed by both platforms?
common <- intersect(limma.filter.array.results$gene.id, voom.limma.deep.results$gene.id)

# extract the common genes
array.common <- subset(limma.filter.array.results$q.value, 
                       limma.filter.array.results$gene.id %in% common)
voomlimma.common <- subset(voom.limma.deep.results$q.value, 
                           voom.limma.deep.results$gene.id %in% common)

# make data frame for plotting
common.df <- data.frame(platform=c(rep("array", each=length(array.common)), 
                                   rep("voomlimma", each=length(voomlimma.common))),
                        q.value=c(array.common, voomlimma.common))

# plot
ggplot(common.df, aes(q.value, fill=platform)) +
  geom_density(alpha=0.3) +
  scale_x_log10() +
  theme_bw() +
  xlab("log(q.value)") +
  ggtitle("q values of common genes analyzed by both platforms")
```

Generate density plot that represents q.values of all genes analyzed by both platforms.
```{r}
# data frame of q-values for all genes 
all.df <- data.frame(platform=c(rep("array", each=nrow(limma.filter.array.results)),
                                   rep("voomlimma", each=nrow(voom.limma.deep.results))),
                        q.value=c(limma.filter.array.results$q.value, 
                                  voom.limma.deep.results$q.value))
                      
# plot
ggplot(all.df, aes(q.value, fill=platform)) +
  geom_density(alpha=0.3) +
  theme_bw() +
  scale_x_log10() +
  xlab("log(q.value)") +
  ggtitle("q values of all genes analyzed by both platforms")
```

> Make some observations about the strengths of these two platforms.

Based on the density plots, we can see that the `voom+limma` analysis resulted in more lower q values compared to the array analysis. This is the reason why there are more DE genes at FDR < 1e-05 from the `voom+limma` DEA results in comparison to the array analysis. This also means that the `voom+limma` platform is more sensitive to calling DE hits compared to the array analysis.


### Q3b) (2 points) Plots

> Plot the gene expression (i.e., from array data) and the logged counts (i.e., from RNA-Seq data) of: two interesting genes identified as DE by both analyses; one DE gene identified as DE only in the array analysis; one DE gene only in the `voom+limma` analysis; one boring gene in both analyses (i.e., 5 genes total measured with 2 platforms)

Let's start by extracting data for two interesting genes identified as DE by both analyses.
```{r, warning=FALSE}
# pick two DE/interesting genes common in both platforms
common <- intersect(array.dea.fdr$gene.id, 
                    deep.dea.fdr$gene.id)

# extract from the DEA array results because it has both probe IDs and gene IDs
common <- subset(limma.filter.array.results,
                 limma.filter.array.results$gene.id %in% common)

# get the gene.id and probe.id common DE genes
(int.common.genes <- common$gene.id[1:2])
(int.common.probes <- common$probe.id[1:2])

# get gene expression and count data for common DE/interesting genes
int.common <- get.genes(int.common.probes, int.common.genes, x="DE.by.both.analyses")
```

Pick one DE gene identified as DE only in the array analysis.
```{r}
# get DE genes identified only in array analysis
array.only <- setdiff(array.dea.fdr$gene.id, 
                      deep.dea.fdr$gene.id)

# extract from the DEA array results 
array.only <- subset(limma.filter.array.results,
                 limma.filter.array.results$gene.id %in% array.only)

# get the gene.id and probe.id common DE genes
(int.array.gene <- array.only$gene.id[1])
(int.array.probe <- array.only$probe.id[1])

# get gene expression and count data for common DE/interesting genes
int.array.only <- get.genes(int.array.probe, int.array.gene, x="DE.by.array.only")
```

Pick one DE gene only in the `voom+limma` analysis.
```{r}
# get DE genes identified only in voom+limma analysis
voomlimma.only <- setdiff(deep.dea.fdr$gene.id, 
                          array.dea.fdr$gene.id)

# extract from the DEA voom+limma results
voomlimma.only <- subset(voom.limma.deep.results,
                 voom.limma.deep.results$gene.id %in% voomlimma.only)

# get the gene.id and probe.id common DE genes
(int.voomlimma.gene <- voomlimma.only$gene.id[3])
(int.voomlimma.probe <- subset(limma.filter.array.results$probe.id, 
                               limma.filter.array.results$gene.id 
                               %in% int.voomlimma.gene))

# get gene expression and count data for common DE/interesting genes
int.voomlimma.only <- get.genes(int.voomlimma.probe, 
                                int.voomlimma.gene, x="DE.by.voomlimma.only")
```

Pick one boring gene in both analyses. 
```{r}
# pick 100 boring genes from array analysis
bor.array <- tail(limma.filter.array.results, 100)

# pick 100 boring genes from RNA-seq analysis
bor.voomlimma <- tail(voom.limma.deep.results, 100)

# what are the common boring genes? 
(common <- intersect(bor.array$gene.id, bor.voomlimma$gene.id))

# pick one boring gene from the list
(common <- common[1])

# extract from the DEA array results because it has both probe IDs and gene IDs
common <- subset(limma.filter.array.results,
                 limma.filter.array.results$gene.id %in% common)

# get the gene.id and probe.id common DE genes
(bor.common.gene <- common$gene.id)
(bor.common.probe <- common$probe.id)

# get gene expression and count data for common DE/interesting genes
bor.common <- get.genes(bor.common.probe, bor.common.gene, x="not.DE.by.both.analyses")
```

Now we can create the big data frame with all the selected genes and plot gene expression/read counts vs. growth condition. 
```{r}
# create big data frame of interesting and boring genes
int.bor.df <- rbind(int.common, int.array.only, int.voomlimma.only, bor.common)

# make ggplot stop reordering my genes
int.bor.df$gene.id <- as.character(int.bor.df$gene.id)
int.bor.df$gene.id <- factor(int.bor.df$gene.id, levels=unique(int.bor.df$gene.id))

# plot 
ggplot(int.bor.df, aes(x=sample, y=value, color=DEA.status)) +
  geom_point() +
  facet_grid(DEA ~ gene.id, scales="free") +
  stat_summary(aes(group=1), fun.y=mean, geom="line") +
  theme_bw() +
  theme(axis.text.x=element_text(angle = -90, hjust = 0)) +
  xlab("Growth condition") +
  ylab("Count/Gene expression") 
```


## Q4: Deep vs low sequencing

### Q4a) (1 point) `voom+limma` DEA of low sequencing data

> In Q2b you analyzed deep count data using `voom+limma`. You will now repeat the analysis on the low sequencing data.

> i) Load the low depth count data using `read.table`; you will need to pass the arguments `header=TRUE` and `row.names=1`. The data file is available as stampy.low.counts.tsv.

```{r}
# load low depth count data
low.count.dat <- read.table("../../data/stampy.low.counts.tsv", header=TRUE, row.names=1)

# smell test
str(low.count.dat)
```

The low counts dataset has 7126 rows and 6 columns. 

> ii) Repeat Q2b-i and Q2b-ii for this new data.

> Package these results in a data.frame called 'voom.limma.low.results' with five columns as you did in Q2b.

Repeat Q2b(i).
```{r}
# normalize counts
norm.factor <- calcNormFactors(low.count.dat)
```

Repeat Q2b(ii).
```{r, message=FALSE}
# convert count data to logged CPM using voom()
low.count.norm <- voom(low.count.dat, design, plot=TRUE,
                  lib.size=colSums(low.count.dat)*norm.factor)

# perform DEA using limma and package the data as specified
voom.limma.low.results <- do.limma.dea(low.count.norm, array=FALSE)

# number of genes with FDR < 1e-05
nrow(subset(voom.limma.low.results, q.value < 1e-05))
```

### Q4b) (2 points) `edgeR` DEA of deep sequencing data

> Now you will use `edgeR` to identify differentially expressed genes between the batch medium vs. chemostat conditions in the deep count datasets loaded in Q2a.

> i) Recall that `edgeR` needs to estimate the dispersion parameter in the negative binomial model using an empirical Bayes method. Estimate the dispersion parameters using `estimateGLMCommonDisp`, `estimateGLMTrendedDisp` and `estimateGLMTagwiseDisp`.

> ii) Use the `glm` functionality of `edgeR`, i.e. use the `glmFit` function, to identify differentially expressed genes between conditions.

> Package these results in a data.frame called 'edger.deep.results' with five columns:
  - gene.id - The id of the gene which reads were aligned to.
  - p.value - The raw p-value for the gene.
  - q.value - The BH corrected p-value, aka the q-value.
  - log.fc - The log fold change which is the column called "logFC" in the edgeR results table.
  - test.stat - The test statistic, which for edgeR is a likelihood ratio. This is the column called "LR" in the edgeR results table.

Perform all the above using the `do.edgeR.dea()` function.
```{r, message=FALSE}
edger.deep.results <- do.edgeR.dea(deep.count.dat)
```

> iii) How many genes are differentially expressed between conditions at a false discovery rate (FDR) of 1e-05? Compare the results with those obtained in Q2b-iii.

```{r}
# get number of DE genes at FDR < 1e-05
nrow(edger.deep.results[edger.deep.results$q.value < 1e-05,])
```

There are `r nrow(edger.deep.results[edger.deep.results$q.value < 1e-05,])` DE genes identified using `edgeR` analysis compared to `r nrow(subset(voom.limma.deep.results, q.value < 1e-05))` DE genes identified using `voom+limma`.

### Q4c) (1 point) edgeR DEA of low sequencing data

> Repeat Q4b-i and Q4b-ii for the low count data. 

> Package these results in a data.frame called 'edger.low.results' with five columns as before.

```{r, message=FALSE}
edger.low.results <- do.edgeR.dea(low.count.dat)

# get number of DE genes at FDR < 1e-05
nrow(edger.low.results[edger.low.results$q.value < 1e-05,])
```

For the low count data, there are `r nrow(edger.low.results[edger.low.results$q.value < 1e-05,])` DE genes identified using `edgeR` analysis compared to `r nrow(subset(voom.limma.low.results, q.value < 1e-05))` DE genes identified using `voom+limma`.

### Q4d) (2 point) Comparison of DEA

> Now that we have the results of the differential expression analysis performed by `voom+limma` and `edgeR` methods on both low and deep count data, we are going to compare and illustrate the results.

> Create a Venn diagram showing all genes identified as differentially expressed (at FDR of 1e-05) in the four previous RNA-Seq analyses. If the comparison of 4 sets gets very confusing, you can also create different pairs of Venn diagrams of interest.

```{r}
# basically used the codes given in the homework
# extract DE genes at FDR < 1e-05
voomlimma.deep <- subset(voom.limma.deep.results$gene.id, 
                          voom.limma.deep.results$q.value < 1e-05)

voomlimma.low <- subset(voom.limma.low.results$gene.id, 
                         voom.limma.low.results$q.value < 1e-05)

edger.deep <- subset(edger.deep.results$gene.id, 
                     edger.deep.results$q.value < 1e-05)

edger.low <- subset(edger.low.results$gene.id, 
                     edger.low.results$q.value < 1e-05)

# assign names to the lists for plotting
de.genes <- list(voomlimma.deep=voomlimma.deep, voomlimma.low=voomlimma.low,
                 edger.deep=edger.deep, edger.low=edger.low)

# start a new plot
plot.new()

# draw the Venn diagram
venn.plot <- venn.diagram(de.genes, filename = NULL, 
                          fill = c("red", "blue", "green", "yellow"))

# draw the plot on the screen
grid.draw(venn.plot)
```

> i) How many genes were identified by voom+limma in both low and deep count data?

None.

> ii) How many genes were identified by edgeR in both low and deep count data?

`r 478+7` genes. 

> iii) How many genes were identified in all the analyses?

None. This is because no DE genes were identified at FDR < 1e-05 using the `voom+limma` analysis for the low counts data.

> iv) Comment on the effect of sequencing depth on the DEA results. Is one of the methods more robust to this effect than the other? Make any additional observations about your results that you find interesting.

When sequencing depth is reduced, DE genes identified by the `voom+limma` method reduced from 1794 genes to none. As for the `edgeR` method, reduced sequencing depth resulted in a decrease in DE genes from 2669 to 486. 

The `edgeR` method is more robust to reduced sequencing depth because `r round(485/2669*100)`% of DE genes identified in the DEA of the deep sequencing depth dataset were retained in the DEA results of the low sequencing depth dataset. In contrast to the `edgeR` method, no DE genes were retained using the `voom+limma` method when sequencing depth was reduced. 

