---
title: "hw1"
author: "Eva Y"
date: "February 14, 2015"
output:
  html_document:
    keep_md: yes
---

### Q0 (0 pts) Intake

**Attach packages for this assignment:**
```{r}
library(car)
library(devtools)
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(gplots))
library(ggplot2)
library(limma)
library(RColorBrewer)
library(reshape2)
```

**Load data and metadata.**
```{r}
# load data
data <- read.delim("~/Documents/stat540_seminar_git/data/data.txt", header = TRUE)

# smell test, what 'flavor' is each variable?
str(data)

# load metadata
metadata <- read.delim("~/Documents/stat540_seminar_git/data/design.txt", header = TRUE)

# smell test, what 'flavor' is each variable?
str(metadata)
```

From this, we can tell that there are `r nrow(data)` genes evaluated by the high-density oligonucleotide microarray and `r ncol(data)` samples. All gene expression are numeric. The metadata provides us with information on the `Treatment` and `time` of the `r ncol(data)` samples. `Treatment` and `time` are factor variables and consist of 2 levels and 4 levels respectively. 

**Let's explore the levels of these factors.**
```{r}
# what are the levels of Treatment? 
levels(metadata$Treatment)

# what are the levels of time?
levels(metadata$time)

# 'change' the levels for time
metadata$time <- factor(metadata$time, levels = c("1_h","2_h", "4_h", "24_h"))
```

**Bugs me a little that 't' in time is lowercase.**
```{r}
# change it to lower case
colnames(metadata)[3] <- "Time"
```

**Perform some basic data cleaning.**
```{r}
# let's change the really long sample IDs for both data and metadata to ease future data exploration and analysis
change_sampleID <- function(d){sub("GSE10718_Biomat_", "", d)}

colnames(data) <- change_sampleID(colnames(data))
row.names(metadata) <- change_sampleID(row.names(metadata))

# let's check if they match up
identical(colnames(data), row.names(metadata)) # looks great!
```

**Now, we are ready to explore the dataset.**


### Q1 (1 points) What are the basic characteristics of the data and meta-data?

> #### Q1a: How many probes? How many samples (Bioassay)?

```{r}
nrow(data) # number of probes
ncol(data) # number of samples
```

There are `r nrow(data)` probes and `r ncol(data)` samples.

> #### Q1b: What is the breakdown of samples (Bioassay) for agent, time?

```{r}
addmargins(with(metadata, table(Treatment, Time)))
```

There are 4 time points namely 1, 2, 4, and 24 hours. For each time-point, there are three untreated samples and three samples treated with cigarette smoke, except for the 1 hour time point which only has two samples.

> #### Q1c: Create a quantitative (numeric) variable that represents the time at which cells were measured.

```{r}
# change factor to numeric
metadata$Time_Hours <- recode(metadata$Time, "'1_h' = 1; '2_h' = 2; '24_h' = 24; '4_h' = 4", 
                              as.factor.result = FALSE)

# check if the level is numeric
str(metadata$Time_Hours) 
```

> #### Q1d: Create a plot showing the gene expression data for one probe and the averages for all possible combinations of agent and time.

```{r}
# pick a random gene and create data frame
set.seed(1)
random_gene <- sample(1:nrow(data), size = 1)
random_gene <- data[random_gene, ]
Gene_Exp <- t(random_gene)

# add sample Treatment and Time
random_gene_df <- cbind(Gene_Exp, metadata[2:3])
  
# make dataset tall and skinny using melt() from reshape2 package
random_gene_df <- melt(random_gene_df, 
                       id.vars=c("Treatment", "Time"), 
                       variable.name="Gene", 
                       value.name="Gene_Exp")
  
# plot using ggplot2
ggplot(random_gene_df, aes(x=Time, y = Gene_Exp, color=Treatment)) + 
  geom_point() + 
  facet_wrap(~Gene) + 
  stat_summary(aes(group=Treatment), fun.y=mean, geom="line") +
  xlab("Time post-treatment") + 
  ylab("Gene expression")
```


### Q2 (2 points) Assessing data quality

> #### Q2a: Examine the sample-to-sample correlations in a heatmap.

**Let's order the samples by time; and within each time group, sort by treatment.**
```{r}
# prepare data frame
# transpose data and create data frame 
t_data <- data.frame(t(data))

# add Treatment and Time columns
t_data <- cbind(metadata$Treatment, metadata$Time_Hours, t_data)

# rename columns
colnames(t_data)[1:2] <- c("Treatment", "Time_Hours")

# order the samples by time; within each time group, sort on treatment
time_treatment <- t_data[order(t_data$Time_Hours, t_data$Treatment), ]

# make ColSideColors and RowSideColors for the heatmap
# for Treatment labels
sample_treatment <- data.frame(time_treatment[ ,1])
colnames(sample_treatment) <- "Treatment"
sample_treatment$Treatment <- gsub("cigarette_smoke", "darkorchid", sample_treatment$Treatment)
sample_treatment$Treatment <- gsub("control", "darkred", sample_treatment$Treatment)

# for Time_Hours labels
sample_time <- data.frame(time_treatment[ ,2])
colnames(sample_time) <- "Time_Hours"
sample_time$Time_Hours[sample_time$Time_Hours == "1"] <- "red"
sample_time$Time_Hours[sample_time$Time_Hours == "2"] <- "blue"
sample_time$Time_Hours[sample_time$Time_Hours == '4'] <- "yellow"
sample_time$Time_Hours[sample_time$Time_Hours == '24'] <- "green"

# remove Treatment and Time_Hours columns
drops <- c("Treatment","Time_Hours")
time_treatment <- time_treatment[ ,!(names(time_treatment) %in% drops)]

# create sample-sample correlation matrices
cmat_time_treatment <- cor(t(time_treatment))

# create a white-blue palette
blu <- colorRampPalette(brewer.pal(n = 9, "Blues"))
palette_size <- 256
blu_palette <- blu(palette_size)

# plot with heatmap.2
heatmap.2(cmat_time_treatment, col=blu_palette, 
          dendrogram="none", Rowv=FALSE, Colv=FALSE, 
          trace="none", ColSideColors=t(sample_time), 
          RowSideColors=t(sample_treatment), 
          main="Samples ordered by treatment , then time")

# add legend
legend("top", legend=c("1 hr","2 hrs","4 hrs","24 hrs","Cigarette smoke","Control"),
       fill=c("red","blue","yellow","green","darkorchid","darkred"), 
       border=FALSE, bty="n", y.intersp=0.7, cex=0.7, ncol=3)
```

**Now, order the samples by treatment; and within each treatment group, sort by time.**
```{r}
# order the samples by treatment; within each treatment group, sort on time
treatment_time <- t_data[order(t_data$Treatment, t_data$Time_Hours), ]

# make ColSideColors and RowSideColors for the heatmap
# for Treatment labels
sample_treatment2 <- data.frame(treatment_time[ ,1])
colnames(sample_treatment2) <- "Treatment"
sample_treatment2$Treatment <- gsub("cigarette_smoke", "darkorchid", sample_treatment2$Treatment)
sample_treatment2$Treatment <- gsub("control", "darkred", sample_treatment2$Treatment)

# for Time_Hours labels
sample_time2 <- data.frame(treatment_time[ ,2])
colnames(sample_time2) <- "Time_Hours"
sample_time2$Time_Hours[sample_time2$Time_Hours == "1"] <- "red"
sample_time2$Time_Hours[sample_time2$Time_Hours == "2"] <- "blue"
sample_time2$Time_Hours[sample_time2$Time_Hours == '4'] <- "yellow"
sample_time2$Time_Hours[sample_time2$Time_Hours == '24'] <- "green"

# remove Treatment and Time_Hours columns
drops <- c("Treatment","Time_Hours")
treatment_time <- treatment_time[ ,!(names(treatment_time) %in% drops)]

# create sample-sample correlation matrices
cmat_treatment_time <- cor(t(treatment_time))

# plot with heatmap.2
heatmap.2(cmat_treatment_time, col=blu_palette, 
          dendrogram="none", Rowv=FALSE, Colv=FALSE, 
          trace="none", ColSideColors=t(sample_time2), 
          RowSideColors=t(sample_treatment2), 
          main="Samples ordered by treatment , then time")

# add legend
legend("top", legend=c("1 hr","2 hrs","4 hrs","24 hrs","Cigarette smoke","Control"),
       fill=c("red","blue","yellow","green","darkorchid","darkred"), 
       border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, ncol=3)
```

**I got a bit annoyed with interpreting the heatmaps generated using `heatmap.2`. `heatmap.2` wouldn't allow me to feed a matrix to the `ColSideColors` argument, which is helpful for this dataset as we are ordering it by two variables - treatment and time. So, I tried `heatmap.3` and it turned out to be quite nice.**

```{r}
# need packages gplots and devtools

# source the function from this url
source_url("https://raw.githubusercontent.com/obigriffith/biostar-tutorials/master/Heatmaps/heatmap.3.R")

# prepare data frame
t_data2 <- t(data)
t_data2 <- cbind(metadata$Treatment, metadata$Time_Hours, t_data2)
colnames(t_data2)[1:2] <- c("Treatment", "Time_Hours")

# order the samples by time; within each time group, sort on treatment
o_data <- t_data2[order(t_data2[ ,2], t_data2[ ,1]), ]
o_data <- data.frame(o_data)

# make ColSideColors matrix
sample_matrix <- data.frame(o_data[ ,1:2])
sample_matrix$Treatment[sample_matrix$Treatment == '1'] <- "darkorchid" # cigarette smoke
sample_matrix$Treatment[sample_matrix$Treatment == '2'] <- "darkred" # control
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '1'] <- "red"
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '2'] <- "blue"
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '4'] <- "yellow"
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '24'] <- "green"
sample_matrix <- as.matrix(sample_matrix)

# get rid of Treatment and Time_Hours columns
drops <- c("Treatment","Time")
time_treatment <- o_data[ ,!(names(o_data) %in% drops)]
cmat_time_treatment <- cor(t(time_treatment))

# plot heatmap
heatmap.3(cmat_time_treatment, scale="none", dendrogram="none",
          Rowv=FALSE, Colv=FALSE, ColSideColors=sample_matrix, 
          symbreaks=FALSE, key=TRUE, symkey=FALSE, 
          density.info="none", trace="none", 
          main="Samples ordered by time , then treatment", 
          cexRow=1, col=rev(grey(seq(0,1,0.01))), 
          ColSideColorsSize=4, 
          KeyValueName="Pearson's Correlation")

# add legend
legend("top", legend=c("1 hr","2 hrs","4 hrs","24 hrs","Cigarette smoke","Control"), 
       fill=c("red","blue","yellow","green","darkorchid","darkred"), 
       border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, ncol=3)
```

**Looks great!**

```{r}
# now try order the samples by treatment; within each treatment group, sort on time
o_data2 <- t_data2[order(t_data2[ ,1], t_data2[ ,2]), ]
o_data2 <- data.frame(o_data2)

# get rid of Treatment and Time_Hours columns
drops <- c("Treatment","Time_Hours")
treatment_time <- o_data2[ ,!(names(o_data2) %in% drops)]
cmat_treatment_time <- cor(t(treatment_time))

# make ColSideColors matrix
sample_matrix <- data.frame(o_data2[ ,1:2])
sample_matrix$Treatment[sample_matrix$Treatment == '1'] <- "darkorchid" # cigarette smoke
sample_matrix$Treatment[sample_matrix$Treatment == '2'] <- "darkred" # control
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '1'] <- "red"
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '2'] <- "blue"
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '4'] <- "yellow"
sample_matrix$Time_Hours[sample_matrix$Time_Hours == '24'] <- "green"
sample_matrix <- as.matrix(sample_matrix)

# plot heatmap
heatmap.3(cmat_treatment_time, scale="none", dendrogram="none",
          Rowv=FALSE, Colv=FALSE, ColSideColors=sample_matrix, symbreaks=FALSE, 
          key=TRUE, symkey=FALSE, density.info="none", trace="none", 
          main="Samples ordered by treatment, then time", cexRow=1, 
          col=rev(grey(seq(0,1,0.01))), ColSideColorsSize=4, 
          KeyValueName="Pearson's Correlation")

# add legend
legend("top", legend=c("1 hr","2 hrs","4 hrs","24 hrs","Cigarette smoke","Control"), 
       fill=c("red","blue","yellow","green","darkorchid","darkred"), 
       border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, ncol=3)
```

**What does the sample correlation matrix tell us about the overall impact of time and agent?**

The correlation performed here uses the Pearson's method. The second heatmap in which the samples are ordered by time, then treatment is easier to interpret than the first heatmap in which the samples are ordered by treatment, then time. 

Based on the second heatmap, we can observe 4 different quadrants. The top-left and bottom-right quadrants are correlations between samples of the same treatment group, which are cigarette smoke treated and untreated control respectively. We can see that both quadrants have darker tiles closer to the same-sample correlation tiles (darkest tiles because correlation = 1). This indicates that gene expression are more correlated within each sample group at closer time points. 

As for the top-right and bottom-left quadrants, the correlation between cigarette smoke treated and controls samples are presented. If we focus on the bottom-left quadrant, we can notice that the tiles goes from darker to lighter horizontally. This indicates that as cigarette smoke treated samples are sampled at later times points, gene expression becomes less correlated between treated and untreated samples. Similarly, this 'gradient' can be observed in the top-right quadrant from top to bottom. 

The first heatmap is harder to interpret and this could be due to the impact of treatment (cigarette smoke treatment vs. control) having a greater effect on gene expression compared to different time points. It is also noticeable that sample 10 (control, 1-hour time point) seems to be an outlier as depicted by lighter tiles (lower correlation) across the heatmap. If we focus on the top-right quandrant, we can tell that sample 10 displays lower correlations with the treated samples in comparison to sample 11 and 12 which are similarly treated and assayed at the 1-hour time point.  

> #### Q2b: Assess the presence of outlier samples.

As described in the heatmap intepretation above, sample 10 seems to be an outlier.

**Now, let's quantify and plot this!**

We can determine an outlier by identifying mean correlation coefficient that fall out of the mean +/- standard deviation.
Data visualization with a  boxplot should do the trick!
```{r}
# create data frame for plotting
cmat_tall <- melt(cmat_time_treatment)

 # remove correlations between same samples because those values are not that informative
rm_same <- which(with(cmat_tall, value == 1))

# sanity check: there should be 23 correlations between same samples
length(rm_same)

# removed, new data frame without correlations between same samples
cmat_tall <- cmat_tall[-rm_same, ]

# determine mean and sd 
summary(cmat_tall$value)
cmat_mean <- mean(cmat_tall$value)
cmat_sd_p <- cmat_mean + sd(cmat_tall$value) # mean plus sd
cmat_sd_n <- cmat_mean - sd(cmat_tall$value) # mean minus sd

# make sure ggplot2 don't reorder my sample sequence, keep it ordered by time followed by treatment
# turn sample(column = Var2) into character vector
cmat_tall$Var2 <- as.character(cmat_tall$Var2) 
cmat_tall$Var2 <- factor(cmat_tall$Var2, levels=unique(cmat_tall$Var2))

# plot boxplot
ggplot(cmat_tall, aes(x = Var2, y = value)) + geom_boxplot() + 
  geom_hline(aes(yintercept = cmat_mean, colour = "red")) + 
  geom_hline(aes(yintercept=cmat_sd_p, colour = "blue")) + 
  geom_hline(aes(yintercept=cmat_sd_n, colour = "blue")) + 
  xlab("Samples") + ylab("Correlation coefficient") + 
  ggtitle("Sample-to-sample correlation (order by time, then treatment)")
```

The blue line in the boxplot is the mean of the correlation matrix whereas the red lines are the mean +/- standard deviation. From the boxplot, we can see that the mean of sample 10 is the only one out of the range of mean of the correlation matrix +/- standard deviation. This indicates that sample 10 is an outlier within this correlation matrix.

**Another method would be to quantify number of correlation coefficients for each sample that fall outside of the mean +/- standard deviation range.**
```{r}
# extract samples with correlation coefficients that are out of range which are considered outliers 
cmat_outlier <- subset(cmat_tall, value > cmat_sd_p | value < cmat_sd_n)
cmat_outlier <- as.data.frame(table(cmat_outlier$Var2))

# pick the sample with the highest frequency of values out of the range
subset(cmat_outlier, Freq == max(cmat_outlier$Freq))
```

From the above analysis, we can see that sample 10 can be confirmed as an outlier with 14 values out of range. 

**Now that sample 10 has been quantified to be an outlier, let's see if it correlate with controls better than other samples treated with cigarette smoke.**
```{r}
# extract data from cmat_tall for sample 10 and correlations with control samples
# use summary() on correlation of sample 10 with control samples
ctrl <- which(with(metadata, Treatment == "control"))
ctrl <- row.names(metadata[ctrl, ])
ctrl <- ctrl[ctrl != "10"]
select_ctrl <- (cmat_tall$Var1 %in% ctrl & cmat_tall$Var2 == "10")
corr_ctrl <- cmat_tall[select_ctrl, ]
summary(corr_ctrl$value)

# extract data from cmat_tall for sample 10 and correlations with cigarette smoke treated samples
# use summary() on correlation of sample 10 with cigarette smoke treated samples 
treated <- which(with(metadata, Treatment == "cigarette_smoke"))
treated <- row.names(metadata[treated, ])
treated <- treated [treated  != "10"]
select_treated <- (cmat_tall$Var1 %in% treated  & cmat_tall$Var2 == "10")
corr_treated <- cmat_tall[select_treated, ]
summary(corr_treated$value)
```

Despite sample 10 being an outlier, its correlation with its own sample group (control) is higher than its correlation with cigarette smoke treated samples. This can be concluded based on a slightly higher mean Pearson's correlation between sample 10 and control samples (0.9015) versus sample 10 and treated samples (0.8939).

> #### Q2c: Assess the distribution of expression values, separated by agent.

**The sample-sample correlation matrix does not capture differences between pairs of samples that are caused by systematic up- or down-regulation of all/most genes. Can you explain why?**

The Pearson's correlation measures the strength of linear relationship between two variables by fitting a best-fit line. In this case, the Pearson's correlation tells us how related two samples are to each other by how strongly the best-fit line fits the relationship (r closer to 1/-1, more correlated). If most genes are systematically up/downregulated in a pair of samples, this moves the best-fit line up or down but does not change how the best-fit line is fitted. Hence, this difference is not captured by Pearson's correlation.

Some help from these links: 
- [link1](https://statistics.laerd.com/statistical-guides/pearson-correlation-coefficient-statistical-guide.php)
- [link2](http://mathworld.wolfram.com/CorrelationCoefficient.html)

**Let's determine if there is a "shift" in distribution of expression levels by plotting two histograms and comparing range, mean, and median.**
```{r}
# prepare tall and skinny data frame for plotting
h_data <- melt(t_data, id.vars = c("Treatment", "Time_Hours"))

# plot histogram
ggplot(h_data, aes(x=value)) + geom_histogram(binwidth=.5, colour="black", fill="white") + 
  facet_wrap(~Treatment) + xlab("Gene expression") + 
  ggtitle("The effects of cigarette smoke on gene expression in NHBE cells") + 
  geom_vline(aes(xintercept=mean(value)), color="red", linetype="dashed", size=1) + 
  geom_vline(aes(xintercept=median(value)), color="blue", linetype="dashed", size=1)
```

The red line is the mean and the blue line is the median. The values seem to be similar for the control and cigarette smoke treated group. Shape and range of both histograms look similar indicating there is no shift. We can also tell that both sets of gene expression data are skewed to the left and this skewness results in median being smaller than mean for both sets of data. 


### Q3 (4 points) Assess differential expression with respect to treatment.
> #### Q3a: Fit a linear model, modeling expression level of each probe using treatment as a single covariate.

**Write out in English and as an equation the model you are using to assess differential expression. In the context of that model, what statistical test are you performing?**

The linear model used to assess differential expression between control and cigarette smoke treated samples is **Y = Xa + b + e**, where Y is the response variable, X is the design matrix of the covariate information (in this case, control=0 and cigarette_smoke=1), a is the coefficient for treatment with cigarette smoke, b is the intercept which is the reference in this model (control group), and e is the error. 

The statistical test performed is a one-way ANOVA.

Some help from this [link](http://www.stat.yale.edu/Courses/1997-98/101/linreg.htm) and STAT540 lecture 7, slide 16. 

**This is how the linear model is fitted...**
```{r}
# arrange the factor levels for Treatment
metadata$Treatment <- factor(metadata$Treatment, levels = c("control", "cigarette_smoke"))

# make design matrix, ref + treatment approach
treatment_des <- model.matrix(~Treatment, metadata)
str(treatment_des)

# fit the linear model
treatment_fit <- lmFit(data, treatment_des)

# apply eBayes() to moderate the estimated error variances
treatment_ebfit <- eBayes(treatment_fit)

# differentially expressed genes between control and cigarette smoke treated
diff_exp_hits <- topTable(treatment_ebfit)

# plot and check out the top 10 hits
top_hits <- row.names(diff_exp_hits)
top_hits <- data.frame(t(data[top_hits, ]))
top_hits <- cbind(metadata$Treatment, top_hits)
colnames(top_hits)[1] <- "Treatment"
colnames(top_hits) <- sub("X", "", colnames(top_hits))
top_hits_tall <- melt(top_hits, id.vars = "Treatment", variable.name = "gene", value.name = "gene_exp")

ggplot(top_hits_tall, aes(x=Treatment, y=gene_exp)) + 
  geom_point() + facet_wrap(~gene) + ylab("Gene expression") +
  stat_summary(aes(group=1), fun.y=mean, geom="line")
```

> #### Q3b: Count your hits, and explore them.

**How many hits (probes) are associated with treatment at unadjusted p-value < 1e-03? How may are associated with treatment at FDR < 0.05?**
```{r}
all_genes_treatment <- topTable(treatment_ebfit, number=nrow(treatment_ebfit), adjust.method = "fdr")

nrow(subset(all_genes_treatment, P.Value < 1e-03))
nrow(subset(all_genes_treatment, adj.P.Val < 0.05))
```

There are `r nrow(subset(all_genes_treatment, P.Value <1e-03))` hits associated with treatment at unadjusted p-value < 1e-03 and `r nrow(subset(all_genes_treatment, adj.P.Val <0.05))` associated with treatment at FDR < 0.05. 

**Take the top 50 probes as your “hits” and create a heatmap of their expression levels. Sort the hits by p-values and the samples by treatment.**
```{r}
# select top 50 hits
top_fifty_hits_df <- all_genes_treatment[1:50, ]
top_fifty_hits <- row.names(top_fifty_hits_df)

top_fifty_hits <- data.frame(t(data[top_fifty_hits, ]))
top_fifty_hits <- cbind(metadata$Treatment, top_fifty_hits)
colnames(top_fifty_hits)[1] <- "Treatment"
colnames(top_fifty_hits) <- sub("X", "", colnames(top_fifty_hits))
top_fifty_hits <- top_fifty_hits[order(top_fifty_hits$Treatment, decreasing = TRUE), ]

# get sample order so we can add ColSizeColors
sample_order <- data.frame(top_fifty_hits$Treatment)
colnames(sample_order) <- "Treatment"
sample_order$Treatment <- gsub("cigarette_smoke", "darkorchid", sample_order$Treatment)
sample_order$Treatment <- gsub("control", "darkred", sample_order$Treatment)

# drop the Treatment column
drops <- c("Treatment")
top_fifty_hits_hmap <- as.matrix(top_fifty_hits[ ,!(names(top_fifty_hits) == "Treatment")])

# plot heatmap
heatmap.2(t(top_fifty_hits_hmap), col=blu_palette, dendrogram = "none", 
          Rowv=FALSE, Colv=FALSE, trace="none", main="Top 50 hits", 
          ColSideColors=t(sample_order))

# add legend
legend("topright", legend=c("Cigarette smoke","Control"), 
       fill=c("darkorchid","darkred"), border=FALSE, 
       bty="n", y.intersp = 0.7, cex=0.7)
```

**What is the (estimated) false discovery rate of this "hits" list? How many of these hits do we expect to be false discoveries?**
```{r}
top_fifty_hits_fdr <- subset(top_fifty_hits_df, adj.P.Val == max(adj.P.Val))
```

The maximum false discovery rate is `r unique(top_fifty_hits_fdr$adj.P.Val)`. Therefore, the number of hits we expect to be false discoveries is `r round(unique(top_fifty_hits_fdr$adj.P.Val)*50)`.

> #### Q3c: Plot the expression levels for a few top (interesting) probes, and a few non-associated (boring) probes.

**Here are some interesting genes!**
```{r}  
# pick top 6 genes and bottom 6 genes using head() and tail()
int_genes <- row.names(head(all_genes_treatment))
bor_genes <- row.names(tail(all_genes_treatment))

# create data frames for interesting genes
int_genes <- data.frame(t(data[int_genes, ]))
int_genes <- cbind(metadata$Treatment, int_genes)
colnames(int_genes)[1] <- "Treatment"
colnames(int_genes) <- sub("X", "", colnames(int_genes))
int_genes_tall <- melt(int_genes, id.vars = "Treatment", variable.name = "gene", value.name = "gene_exp")

# plot for interesting genes
ggplot(int_genes_tall, aes(x = Treatment, y = gene_exp)) + 
  geom_point() + facet_wrap(~gene) + ylab("Gene expression") +
  stat_summary(aes(group=1), fun.y=mean, geom="line")
```

**Here are some boring ones...**
```{r}
# create data frames for boring genes
bor_genes <- data.frame(t(data[bor_genes, ]))
bor_genes <- cbind(metadata$Treatment, bor_genes)
colnames(bor_genes)[1] <- "Treatment"
colnames(bor_genes) <- sub("X", "", colnames(bor_genes))
bor_genes_tall <- melt(bor_genes, id.vars = "Treatment", variable.name = "gene", value.name = "gene_exp")

# plot for interesting genes
ggplot(bor_genes_tall, aes(x = Treatment, y = gene_exp)) + 
  geom_point() + facet_wrap(~gene) + ylab("Gene expression") +
  stat_summary(aes(group=1), fun.y=mean, geom="line")
```


### Q4 (4 points) Assess differential expression with respect to time.
> #### Q4a: Fit a linear model, assessing the effect of time on gene expression

**Now that we're asked to assess time, we need a new design matrix.**
```{r}
# make design matrix, ref + treatment approach
time_des <- model.matrix(~Time_Hours, metadata)
str(time_des)

# fit the linear model
time_fit <- lmFit(data, time_des)

# apply eBayes() to moderate the estimated error variances
time_ebfit <- eBayes(time_fit)

# differentially expressed genes between time points
diff_exp_hits <- topTable(time_ebfit)

# plot and check out the top 10 hits
top_hits <- row.names(diff_exp_hits)
top_hits <- data.frame(t(data[top_hits, ]))
top_hits <- cbind(metadata$Time_Hours, top_hits)
colnames(top_hits)[1] <- "Time_Hours"
colnames(top_hits) <- sub("X", "", colnames(top_hits))
top_hits_tall <- melt(top_hits, id.vars = "Time_Hours", variable.name = "gene", value.name = "gene_exp")

ggplot(top_hits_tall, aes(x=Time_Hours, y=gene_exp)) + 
  geom_point() + facet_wrap(~gene) + ylab("Gene expression") +
  stat_smooth(method = "lm", se = FALSE)
```

**How many hits are associated with time at unadjusted p-value 1e-03? At FDR 0.05?**
```{r}
all_genes_time <- topTable(time_ebfit, number = nrow(time_ebfit), adjust.method = "fdr")

nrow(subset(all_genes_time, P.Value < 1e-03))
nrow(subset(all_genes_time, adj.P.Val < 0.05))
```

There are `r nrow(subset(all_genes_time, P.Value <1e-03))` hits associated with treatment at unadjusted p-value < 1e-03 and `r nrow(subset(all_genes_time, adj.P.Val <0.05))` associated with treatment at FDR < 0.05. 

> #### Q4b: Plot expression levels of a few top probes and a few boring ones:

**Here are some interesting genes!**
```{r}
# pick top 6 genes and bottom 6 genes using head() and tail()
int_genes <- row.names(head(all_genes_time))
bor_genes <- row.names(tail(all_genes_time))

# create data frames for interesting genes
int_genes <- data.frame(t(data[int_genes, ]))
int_genes <- cbind(metadata$Time_Hours, int_genes)
colnames(int_genes)[1] <- "Time_Hours"
colnames(int_genes) <- sub("X", "", colnames(int_genes))
int_genes_tall <- melt(int_genes, id.vars = "Time_Hours", variable.name = "gene", value.name = "gene_exp")

# plot for interesting genes
ggplot(int_genes_tall, aes(x = Time_Hours, y = gene_exp)) + geom_point() + facet_wrap(~gene) + stat_smooth(method = "lm", se = FALSE)
```

**Now, let's plot some boring genes.**
```{r}
# create data frames for boring genes
bor_genes <- data.frame(t(data[bor_genes, ]))
bor_genes <- cbind(metadata$Time_Hours, bor_genes)
colnames(bor_genes)[1] <- "Time_Hours"
colnames(bor_genes) <- sub("X", "", colnames(bor_genes))
bor_genes_tall <- melt(bor_genes, id.vars = "Time_Hours", variable.name = "gene", value.name = "gene_exp")

# plot for boring genes
ggplot(bor_genes_tall, aes(x = Time_Hours, y = gene_exp)) + geom_point() + facet_wrap(~gene) + stat_smooth(method = "lm", se = FALSE)
```

The plots are different from the one in question 1 in which we calculated the **mean** of each **categorical** variable (changes between samples are equal/uniform) and draw a line joining both means. In these plots, the line is a linear regression line fitted by the argument `method = "lm"` in the stat_smooth() function of the `ggplot2` package across hours post treatment which is a **continuous** variable (changes between samples are not equal/uniform and quantitative). 


### Q5 (4 points) Perform differential expression analysis using a full model with both treatment and time as covariates.

> #### Q5a: Quantify the number of hits for treatment.

**Now that we want to take into account both treatment and time as covariates, we need a new design matrix for our model.**
```{r}
# make design matrix
treatment_time_des <- model.matrix(~Treatment*Time_Hours, metadata)
str(treatment_time_des)

# rename columns
colnames(treatment_time_des)[2:4] <- c("cigarette_smoke", "Time_Hours", "cigarette_smoke:Time_Hours")

# fit the linear model
treatment_time_fit <- lmFit(data, treatment_time_des)

# apply eBayes() to moderate the estimated error variances
treatment_time_ebfit <- eBayes(treatment_time_fit)

# differentially expressed genes between treatments or time points or both
diff_exp_hits <- topTable(treatment_time_ebfit)

# plot and check out the top 10 hits
top_hits <- row.names(diff_exp_hits)
top_hits <- data.frame(t(data[top_hits, ]))
top_hits <- cbind(metadata$Treatment, metadata$Time_Hours, top_hits)
colnames(top_hits)[1:2] <- c("Treatment", "Time_Hours")
colnames(top_hits) <- sub("X", "", colnames(top_hits))
top_hits_tall <- melt(top_hits, id.vars = c("Treatment", "Time_Hours"), variable.name = "gene", value.name = "gene_exp")

ggplot(top_hits_tall, aes(x = Time_Hours, y = gene_exp, group = Treatment, color = Treatment)) + 
  geom_point() + facet_wrap(~gene) + 
  stat_smooth(aes(group = Treatment), method = "lm", se = FALSE)
```

**For how many probes is treatment a significant factor at the unadjusted p-value 1e-03, and at FDR < 0.05 level?**
```{r}
all_genes_treatment_time <- topTable(treatment_time_ebfit, coef="cigarette_smoke", number = nrow(treatment_time_ebfit), adjust.method = "fdr")

nrow(subset(all_genes_treatment_time, P.Value < 1e-03))
nrow(subset(all_genes_treatment_time, adj.P.Val < 0.05))
```

There are `r nrow(subset(all_genes_treatment_time, P.Value <1e-03))` hits associated with treatment at unadjusted p-value < 1e-03 and `r nrow(subset(all_genes_treatment_time, adj.P.Val <0.05))` associated with treatment at FDR < 0.05. 

**Is this number different from what you reported in Q3b? Why? Quantify the proportion of overlapping probes among your hits, when using the unadjusted p-value threshold of 1e-03.**

```{r}
treatment_pval <- row.names(subset(all_genes_treatment, P.Value < 1e-03))
treatment_time_pval <- row.names(subset(all_genes_treatment_time, P.Value < 1e-03))

venn(list(treatment_only = treatment_pval, full_model = treatment_time_pval))
```

When the model only assess `Treatment`, there are 805 genes that are significant at the unadjusted p-value threshold of 1e-03. As for the full model, there are 621 genes that are significantly associated with `Treatment`. Out of the 621 genes, 328 genes are hits in the treatment-only model which is a percentage of `r signif(328/621*100, digits=2)`%. 

There are less significant genes associated with `Treatment` in the full model because `Time_Hours` (time post treatment) is taken into account. Moreover, 293 genes were found to be significant in the full model that were not hits in the treatment-only model. Therefore, this indicates that time in addition to treament in this experiment has an impact on gene expression in NHBE cells.

**Plot the distributions of all the p-values for treatment when using both models, i.e., one from the model in Q3 and one from the full model in this question. Compare and comment on the similarity/differences in the shape of the distributions.**
```{r}
# create data frame
pval <- data.frame(all_genes_treatment$P.Value, all_genes_treatment_time$P.Value)
colnames(pval) <- c("treatment_only", "full_model")
pval_tall <- melt(pval, variable.name = "model", value.name = "P.Value")

ggplot(pval_tall, aes(P.Value, color = model)) + geom_density()
```

The shape of the distributions looks similar for both full model and treatment-only model. The only difference is the treatment-only model has a slightly higher count for lower p-values than the full model (the peak in the red line is higher than the blue line). Hence, with more genes with lower p-values, it is no surprise that at the p-value < 1e-03 cut off there are more hits for the treatment-only model in comparison to the full model.

> #### Q5b: Test the null hypothesis that there is no significant interaction between time and treatment.

**Explain in English what you are modeling with this interaction term (what does it represent?).**

The `Treatment` and `Time_Hours` interaction term (~Treatment*Time_Hours) models the effect of both treatment and time on gene expression. This model aims to select genes that are differentially expressed between control and exposure to cigarette smoke as well as over time (1-24 hr).

**For how many probes is the interaction effect significant at the unadjusted p-value < 1e-03, and at FDR < 0.05 level?**
```{r}
all_genes_interact <- topTable(treatment_time_ebfit, coef="cigarette_smoke:Time_Hours", number = nrow(treatment_time_ebfit), adjust.method = "fdr")

nrow(subset(all_genes_interact, P.Value < 1e-03))
nrow(subset(all_genes_interact, adj.P.Val < 0.05))
```

There are `r nrow(subset(all_genes_interact, P.Value < 1e-03))` hits associated with treatment at unadjusted p-value < 1e-03 and `r nrow(subset(all_genes_interact, adj.P.Val <0.05))` associated with treatment at FDR < 0.05.

> #### Q5c: Plot a few probes where the interaction does and does not matter

**Here are some genes where the interaction matter!**
```{r}
# pick top 6 genes and bottom 6 genes using head() and tail()
int_genes <- row.names(head(all_genes_interact))
bor_genes <- row.names(tail(all_genes_interact))

# create data frames for genes in which the interaction is significant
int_genes <- data.frame(t(data[int_genes, ]))
int_genes <- cbind(metadata$Treatment, metadata$Time_Hours, int_genes)
colnames(int_genes)[1:2] <- c("Treatment", "Time_Hours")
colnames(int_genes) <- sub("X", "", colnames(int_genes))
int_genes_tall <- melt(int_genes, id.vars = c("Treatment", "Time_Hours"), variable.name = "gene", value.name = "gene_exp")

# plot for genes in which the interaction is significant
ggplot(int_genes_tall, aes(x = Time_Hours, y = gene_exp, group = Treatment, color = Treatment)) + 
  geom_point() + facet_wrap(~gene) + 
  stat_smooth(method = "lm", se = FALSE)
```

**Here are some genes where the interaction does not matter!**
```{r}
# create data frames for genes in which the interaction is not significant
bor_genes <- data.frame(t(data[bor_genes, ]))
bor_genes <- cbind(metadata$Treatment, metadata$Time_Hours, bor_genes)
colnames(bor_genes)[1:2] <- c("Treatment", "Time_Hours")
colnames(bor_genes) <- sub("X", "", colnames(bor_genes))
bor_genes_tall <- melt(bor_genes, id.vars = c("Treatment", "Time_Hours"), variable.name = "gene", value.name = "gene_exp")

# plot for genes in which the interaction is not significant
ggplot(bor_genes_tall, aes(x=Time_Hours, y=gene_exp, group=Treatment, color=Treatment)) + 
  geom_point() + facet_wrap(~gene) + 
  stat_smooth(method = "lm", se = FALSE)
```

#### Bonus question: consider the limitations of the model you used in Q5, can you think of an assumption underlying the model that is not consistent with the specification of this data?
One of the assumptions of linear regression is that the relationship between a response variable and independent variable (covariate) must be **linear**. As for gene expression over time, this is not always the case as presented in lecture 9. Some genes demonstrate a quadratic effect over time so perhaps we should test the data with a quadratic model and make a comparison. 

Some help from this [link](http://people.duke.edu/~rnau/testing.htm)..