---
title: "sm05"
author: "Eva Y"
date: "February 9, 2015"
output:
  html_document:
    keep_md: yes
---

**Write function to create data frame.**
```{r}
library(lattice)
library(ggplot2) 
library(reshape2)

prDat <- read.table("~/Documents/stat540_seminar_git/seminar/sm03/GSE4051_data.tsv")
str(prDat, max.level = 0)

prDes <- readRDS("~/Documents/stat540_seminar_git/seminar/sm03/GSE4051_design.rds")
str(prDes)

(luckyGenes <- c("1419655_at","1438815_at"))

# function that takes as input the Affymetrix probeset ID(s) and gives as output a data.frame
prepareData <- function(x){
  selected_genes <- t(prDat[x, ])
  pDat <- cbind(selected_genes, prDes)
  pDat <- with(pDat, data.frame(sidChar, sidNum, devStage, gType, gene = factor(rep(c(colnames(pDat[x])), each = nrow(pDat))), gExp = c(selected_genes)))
}

jDat <- prepareData(luckyGenes)
str(jDat)
```

**Write a function to plot stripplot using lattice.**
```{r}
# function to plot stripplot using lattice
makeStripplot <- function (x, ...) {
  stripplot(gExp ~ devStage | gene, x,
          group = gType, jitter.data = TRUE,
          auto.key = TRUE, type = c('p', 'a'), grid = TRUE, ...)
  }

makeStripplot(jDat)
makeStripplot(jDat, pch = 17, cex = 3) # I've got control!
```

**Now, try it with ggplot2.**
```{r}
# function to plot stripplot using ggplot2
makeStripplotGg <- function (x) {
  ggplot(x, aes(x = devStage, y = gExp, color = gType, group = gType)) +
  geom_point() + stat_smooth(se = F) + facet_wrap(~ gene)
  }

makeStripplotGg(jDat)
```

**Do a two-sample t-test to determine a difference in expected gene expression for probeset "1456341_a_at" at developmental stage P2 vs. 4 weeks post-natal (ignoring genotype, i.e. lump the wild types and knockouts together). Let’s assume a common variance in the two groups.**

```{r}
# make data frame using function for probeset 1456341_a_at and plot a stripplot
makeStripplot(newDat <- prepareData("1456341_a_at"))

# use subset to select for devStage P2 and 4_weeks
newDat_p2_4weeks <- subset(newDat, devStage == "P2" | devStage == "4_weeks", 
                    select = c(sidChar:gExp))

# perform two sample t-test
(tt_newDat <- t.test(gExp ~ devStage, newDat_p2_4weeks, var.equal = TRUE))
```

**Let's focus on probeset "1438786_a_at" and fit a linear model with a categorical covariate.**
```{r}
# make data frame using function for probeset 1438786_a_at and plot a stripplot
makeStripplot(mDat <- prepareData("1438786_a_at"))

# fit linear model
mFit <- lm(formula = gExp ~ devStage, data = mDat, subset = gType == "wt")
summary(mFit)
```

**As the "W" shape of the expression profile for “1438786_a_at” means that the expression values for developmental stages P2 and P10 are quite similar, let's test whether the P2 and P10 effects are equal or, equivalently, whether their difference is equal to zero.**
```{r}
# extract parameter estimates for the fitted linear model
coef(mFit)

# construct contrast matrix - one row
contMat <- matrix(c(0, 1, 0, -1, 0), 1)
(obsDiff <- contMat %*% coef(mFit))

# check to see if this really is the observed difference in sample mean for wt mice, P2 vs. P10
(sampMeans <- aggregate(gExp ~ devStage, mDat, FUN = mean,
                        subset = gType == "wt"))
with(sampMeans, gExp[devStage == "P2"] - gExp[devStage == "P10"]) # yes!
```

**Let's try obtain the estimated standard error for our contrast.**
```{r}
# use vcov() to obtain variance-covariance matrix of the parameters estimated in the original model
vcov(mFit)

# take diagnonal elements and sqrt them, should exactly equal to standard errors reported for the original model
sqrt(diag(vcov(mFit)))

summary(mFit)$coefficients[ , "Std. Error"] # yes!
```

**We can form a test statistics by dividing observed effect (difference in estimated coefficient between P2 and P10) and estimated standard error.**
```{r}
(estSe <- contMat %*% vcov(mFit) %*% t(contMat)) # variance-covariance matrix of a contrast

# test statistics 
(testStat <- obsDiff/estSe)

# compute two-sided p value 
2 * pt(abs(testStat), df = df.residual(mFit), lower.tail = FALSE)
```

** As p-value is rather larger, we cannot conclude a significant difference in gene expression of 1438786_a_at in wt mice between devStage P2 and P10.**

##### Let's try fit a linear model with two categorical covariates.
```{r}
# let’s focus on probeset “1448690_at”
makeStripplot(oDat <- prepareData("1448690_at"))
str(oDat)

# fit a linear model with covariates gType and devStage and include their interactions. 
oFitBig <- lm(formula = gExp ~ devStage*gType, data = oDat)
summary(oFitBig)$coef

# fit a linear model with same covariates but omit the interaction
oFitSmall <- lm(formula = gExp ~ devStage + gType, data = oDat)
summary(oFitSmall)$coef
```

**Recall the anova() function can take two fitted models, one nested within the other, and perform this test.**
```{r}
anova(oFitBig, oFitSmall) # With a p-value awfully close to one, we confirm that, no, there is no evidence for interaction in this particular case.
```

**Now, try the analysis with probeset “1429225_at” to see if I can duplicated the results.**
```{r}
makeStripplot(oDat <- prepareData("1429225_at"))

# fit a linear model with covariates gType and devStage and include their interactions. 
oFitBig <- lm(formula = gExp ~ gType*devStage, data = oDat)
summary(oFitBig)$coef

# fit a linear model with same covariates but omit the interaction
oFitSmall <- lm(formula = gExp ~ gType + devStage, data = oDat)
summary(oFitSmall)$coef

anova(oFitBig, oFitSmall) # Not surprisingly, the interaction here is highly statistically significant.
```

### Take-home problems
#### Question 1:
**Estimate the effect of devStage on mDat with a linear model by using lm(). You will want to use the following formula: gExp ~ devStage. What stages are significantly different from 4_weeks? Use summary() to get your results.**

```{r}
# load data
mDat <- read.table("~/Documents/stat540_seminar_git/seminar/sm05/mDat.tsv", header=T, sep= "\t", row.names = 1)

mFit <- lm(formula = gExp ~ devStage, data = mDat)
summary(mFit)
```

**P10 and P2 are significantly different from 4_weeks.**

#### Question 2:
**Run the exact same test on mDat values with genotype “NrlKO”. What are the differences between this test and the last one?**

```{r}
mFit_KO <- lm(formula = gExp ~ devStage, data = mDat, subset = gType == "NrlKO")
summary(mFit_KO)
```

**At stage P2 and P10, gene expression of 1438786_a_at in NrlKO mice are still significantly different from stage 4 weeks but less significant than the previous model.**

#### Question 3:
**Fit a linear model of the expression of 1438786_a_at as a linear model of devStage and gType without interactions. Use the following formula: gExp ~ devStage + gType.**
```{r}
mFit_no_intr <- lm(formula = gExp ~ devStage + gType, data = mDat)
summary(mFit_no_intr)
```

#### Question 4:
**Fit a linear model of the expression of 1438786_a_at as a linear model of devStage and gType with interactions. Use the following formula gExp ~ devStage * gType. Is this model different from the model without interctions?**
```{r}
mFit_w_intr <- lm(formula = gExp ~ devStage * gType, data = mDat)
summary(mFit_w_intr)
```

**devStage becomes less significant when gType interaction is introduced.**

#### Question 5:
**Run an anova comparing the models from questions 3 and 4. Are the models significantly different?**
```{r}
anova(mFit_no_intr, mFit_w_intr)
```

**p-value (= 0.7637) is high so the models are not significantly different.**

#### Question 6:
**We are now going to create a batch effect and control for it in our analysis using a blocking variable. Run the following code block to create a new data set, bDat, with a batch effect.**
```{r}
batches = as.character(rep(c("batch1","batch2"),20)[1:39])
batchEffect = rep(c(0, 4), 20)[1:39]
bDat = mDat
bDat$gExp = bDat$gExp + batchEffect 
summary(lm(gExp ~ devStage * gType, data=bDat))
```

**How has the batch effect changed the result of the test? How is it different from the result of question 4?**

When batch effect is included, expression of 1438786_a_at at P2 and P10 are not significantly different from 4 weeks anymore. 

#### Question 7:
**Fit a new linear model with the formula gExp ~ devStage * gType + batches using data=bDat. How is this similar to the results of question 4? In this case, batches is our blocking variable. By including batches in the model we can estimate the effect of contamination in our data caused by the difference in batches and get back to testing hypotheses we’re really interested in, like the effect of knockouts!**
```{r}
bFit <- lm(formula = gExp ~ devStage * gType + batches, data = bDat)
summary(bFit)
```

**Expression of 1438786_a_at at P2 and P10 becomes significantly different from 4 weeks like the results in question 4.**

#### Question 8:
**Run an anova comparing the two formulae gExp ~ devStage + gType + batches and gExp ~ devStage * gType + batches. How are these results similar to that of question 5?**
```{r}
bFit_no_intr <- lm(formula = gExp ~ devStage + gType + batches, data = bDat)
anova(bFit, bFit_no_intr)
```

**Similar to question 5, p-value (= 0.7601) is high so the models are not significantly different.**

