---
title: "sm06_takehome"
author: "Eva Y"
date: "February 22, 2015"
output:
  html_document:
    keep_md: yes
---

#### Take-home problem: 

##### First, let's load R packages and data for the analysis. 
```{r}
library(limma)
library(ggplot2)

prDat <- read.table("~/Documents/stat540_seminar_git/seminar/sm03/GSE4051_data.tsv")
str(prDat, max.level = 0)

prDes <- readRDS("~/Documents/stat540_seminar_git/seminar/sm03/GSE4051_design.rds")
str(prDes)

# extract data for wild-type mice
wtDes <- subset(prDes, gType == "wt")
str(wtDes)

wtDat <- subset(prDat, select = prDes$gType == "wt")
str(wtDat, max.level = 0)

# make design matrix, accept default "ref + treatment"
wtDesMat <- model.matrix(~ devStage, wtDes)
str(wtDesMat)

# fit the model for all probes at once
# use eBayes() to moderate the estimated error variances
wtFit <- lmFit(wtDat, wtDesMat)
wtEbFit <- eBayes(wtFit)
```

##### (1) See if you can find one or more probes that have some expression changes up to P6 and then hold steady all the way to 4_weeks. Plot the probes in your finding. Hereâ€™s some I found.
```{r}
# edit the contrast matrix 
(cont.matrix_2 <- makeContrasts(P2VsE16 = devStageP2 - Intercept, P6VsP2 = devStageP6 - devStageP2, P10VsP6 = devStageP10 - devStageP6, fourweeksVsP10 = devStage4_weeks - devStageP10, levels = wtDesMat))

wtFitCont_2 <- contrasts.fit(wtFit, cont.matrix_2)
wtEbFitCont_2 <- eBayes(wtFitCont_2)

# pick a cut off, more stringent one
cutoff <- 1e-04
wtResCont_2 <- decideTests(wtEbFitCont_2, p.value = cutoff, method = "global")
summary(wtResCont_2) 

# pick the genes with changes from E16-P6 and no changes from P6-4weeks based on the 'more' stringent cut off
hits7 <- rownames(prDat)[which(wtResCont_2[, "P2VsE16"] != 0 & wtResCont_2[, "P6VsP2"] != 0 & wtResCont_2[, "P10VsP6"] == 0 & wtResCont_2[, "fourweeksVsP10"] == 0)]

# source script from previous exercise
source('~/Documents/stat540_seminar_git/source_scripts/make_df_stripplots.R', echo=TRUE)

# make data frame, select only wildtype data, and plot 
hits7_wt <- subset(prepareData(hits7), gType == "wt")
p <- makeStripplotGg(hits7_wt) + theme(text = element_text(size=8))
suppressMessages(print(p))
```

**Using a more stringent cut off for p-value, I got 10 hits.**

##### (2) Repeat (1) for the genotype NrlKO. Compare the hits between the wt and NrlKO. Are there any common genes shared by these 2 groups of hits, what does the result suggest in terms of the effect of genotype?
```{r}
# extract data for KO mice
koDes <- subset(prDes, gType == "NrlKO")
str(koDes)

koDat <- subset(prDat, select = prDes$gType == "NrlKO")
str(koDat, max.level = 0)

# make design matrix, accept default "ref + treatment"
koDesMat <- model.matrix(~ devStage, koDes)
str(koDesMat)

# fit the model for all probes at once
# use eBayes() to moderate the estimated error variances
koFit <- lmFit(koDat, koDesMat)
koEbFit <- eBayes(koFit)

# edit the contrast matrix 
(cont.matrix_3 <- makeContrasts(P2VsE16 = devStageP2 - Intercept, P6VsP2 = devStageP6 - devStageP2, P10VsP6 = devStageP10 - devStageP6, fourweeksVsP10 = devStage4_weeks - devStageP10, levels = koDesMat))

koFitCont <- contrasts.fit(koFit, cont.matrix_3)
koEbFitCont <- eBayes(koFitCont)

# pick a cut off, more stringent one
cutoff <- 1e-04
koResCont <- decideTests(koEbFitCont, p.value = cutoff, method = "global")
summary(koResCont) 

# pick the genes with changes from E16-P6 and no changes from P6-4weeks based on the 'more' stringent cut off
hits8 <- rownames(prDat)[which(koResCont[, "P2VsE16"] != 0 & koResCont[, "P6VsP2"] != 0 & koResCont[, "P10VsP6"] == 0 & koResCont[, "fourweeksVsP10"] == 0)]

# make data frame, select only KO data, and plot 
# select only 10 hits from hits8
hits8_ko <- subset(prepareData(hits8[1:10]), gType == "NrlKO")
p <- makeStripplotGg(hits8_ko) + theme(text = element_text(size=8))
suppressMessages(print(p))

# check if there are any genes shared between wt and ko mice 
common_hits <- intersect(hits7, hits8)
```

##### There are three probes that are commom in between wild-type and knock-out mice that are differentially expressed from E16-P6 and stablizes from E6-4weeks.
```{r}
# make data frame and plot 
hits9 <- prepareData(common_hits)
p <- makeStripplotGg(hits9) + theme(text = element_text(size=8))
suppressMessages(print(p))
```

